Vivado compatibility fixes

FrancescoCiraolo opened the pull request:
I remove os-lib dependecies, replaced with the Resource system (required to be a full blown Scala library) and changed the dependency style for metals/vscode compatibility

FrancescoCiraolo committed:
Added vivado versions 2024.2 and 2019.2

- 2024.2 tested to bitstream (not run on board)
- 2019.2 not tested

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

FrancescoCiraolo committed:
Changed dependencies declaration for compatibiliy with metals plugin

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

FrancescoCiraolo committed:
[FIX librairy] Vivado.scala

Replace the way to obtain the json to be compliant with Java/Scala resources

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

FrancescoCiraolo committed:
Removed os dependency as not required anymore

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

denishoornaert committed:
Add new vivado version to list

FrancescoCiraolo commented:
Nope, I did not attempt the synthesis yet, if you want I can try right away.

FrancescoCiraolo committed:
Merge branch 'master' into wip/vivado-fixes

FrancescoCiraolo committed:
Added function in Vivado.scala to read the IPVersionMap (to avoid single use object members)

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

denishoornaert commented:
Thanks, I will update the readme for the 2024.2 version :)
Do you use the 2019.2 version for the ZCU102? Maybe you could check for that one. Not sure if we still have a ZCU102 license at TUM...

denishoornaert committed:
Mark 2024.2 as tested

FrancescoCiraolo committed:
Batch fixed to test ZCU102 on Vivado 2019.2

- Fixed missing properties
- Updated ZCU102 version
- Fixed messages printing in TCLScript.scala
- Fixed Synthesis and Implementation version in TCLScript.scala

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

denishoornaert commented on hw/spinal/ultrascaleplus/scripts/TCLFactory.scala:
```diff
@@ -253,7 +263,10 @@ object TCLFactory {
   def createSynthesis(fileset: String, constraintFileset: String): String = {
     val strategy = "\"Vivado Synthesis Defaults\""
-    val flow     = "\"Vivado Synthesis 2022\""
+    val flow     = Vivado.version match {
+      case "2019.2" => "\"Vivado Synthesis 2019\""
+      case _ => "\"Vivado Synthesis 2022\""
+    }
```
Interesting. When I tried, with version 24, I got no complaints :sweat_smile: 
Nice catch! Since the format is fairly straightforward, maybe it can b generated. I will give it a shot and push here!

denishoornaert commented on hw/spinal/zcu102/ZCU102.scala:
```diff
@@ -76,7 +76,7 @@ class ZCU102(
 
   override val board = "zcu102"
-  override val version = "3.4"
+  override val version = "3.3"
```
So my vivado version (24) gave me 3.4. I guess (unsurprisingly) the board version also depends on the vivado version... We should probably implement a version mapping similar to the other IPs (ZynqMPSoC, ...) but let us do that later :wink:

denishoornaert committed:
software: tcl generation: change version encoding to year+revision

denishoornaert commented on hw/spinal/ultrascaleplus/scripts/TCLFactory.scala:
```diff
@@ -253,7 +263,7 @@ object TCLFactory {
 
   def createSynthesis(fileset: String, constraintFileset: String): String = {
     val strategy = "\"Vivado Synthesis Defaults\""
-    val flow     = "\"Vivado Synthesis 2022\""
+    val flow = f"\"Vivado Synthesis ${Vivado.year}\""
```
Switched to this for leaner generation

denishoornaert committed:
documentation: update readme with tabel to better capture tested conbinations

denishoornaert committed:
Update README.md

denishoornaert committed:
software suport: vivado version management: add documentation

denishoornaert committed:
software suport: vivado version management: change how vivado version is stored and can be accessed

FrancescoCiraolo commented:
@denishoornaert I was thinking about reading the list of part versions from tcl instead of going for a handmade map, what do you think?
Clearly, I'm hoping it's possible, but no actual idea 😅

FrancescoCiraolo committed:
Implemented Catalog probing of current Vivado version

- Execute a TCL scripts to list the IPs, Boards and parts
- Collect those information and offer methods to get the IP/Board version without relying on hardcoded map

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

FrancescoCiraolo committed:
Added System.exit(-1) to each of the exceptions in detectVivadoVersion \& checked version in VivadoCatalog construction

Signed-off-by: Francesco Ciraolo <fciraolo@bu.edu>

FrancescoCiraolo commented:
@denishoornaert Added a TCL probing of the Vivado catalog and related collections and methods!
Tested up to bitstream on Vivado 2019.2

Commit:
makefile: add recipe to get bitstream

Commit:
documentation: update vivado compatibility

denishoornaert commented:
LGTM!

I tested with 2023.2 on my side and it seems to work flawlessly :smile:
